File: core/src/main/java/dev/bee/kanjianki/core/study/StudyRating.java
package dev.bee.kanjianki.core.study;

public enum StudyRating {
    AGAIN("again", 0),
    HARD("hard", 1),
    GOOD("good", 2),
    EASY("easy", 3);

    private final String code;
    private final int strength;

    StudyRating(String code, int strength) {
        this.code = code;
        this.strength = strength;
    }

    public String code() {
        return code;
    }

    public boolean strongerThan(StudyRating other) {
        return strength > other.strength;
    }

    public StudyRating cappedAt(StudyRating ceiling) {
        return strongerThan(ceiling) ? ceiling : this;
    }

    public static StudyRating fromCode(String code) {
        if (code == null) {
            return AGAIN;
        }
        for (StudyRating rating : values()) {
            if (rating.code.equals(code)) {
                return rating;
            }
        }
        return AGAIN;
    }
}

File: writing-core/src/main/kotlin/dev/bee/kanjianki/core/study/StudyRating.kt
package dev.bee.kanjianki.core.study

import dev.bee.kanjianki.core.StudyRatings

enum class StudyRating(private val code: String, private val strength: Int) {
    AGAIN(StudyRatings.AGAIN, 0),
    HARD(StudyRatings.HARD, 1),
    GOOD(StudyRatings.GOOD, 2),
    EASY(StudyRatings.EASY, 3),
    ;

    fun code(): String = code

    fun strongerThan(other: StudyRating): Boolean = strength > other.strength

    fun cappedAt(ceiling: StudyRating): StudyRating = if (strongerThan(ceiling)) ceiling else this

    companion object {
        @JvmStatic
        fun fromCode(code: String?): StudyRating {
            if (code == null) {
                return AGAIN
            }
            for (rating in entries) {
                if (rating.code == code) {
                    return rating
                }
            }
            return AGAIN
        }
    }
}
